invenio_webhooks/receivers.py

```diff
@@ -0,0 +1,207 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Invenio.
+# Copyright (C) 2016 CERN.
+#
+# Invenio is free software; you can redistribute it
+# and/or modify it under the terms of the GNU General Public License as
+# published by the Free Software Foundation; either version 2 of the
+# License, or (at your option) any later version.
+#
+# Invenio is distributed in the hope that it will be
+# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Invenio; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+"""Useful webhook receivers."""
+
+from __future__ import absolute_import
+
+
+from celery import chain, group, shared_task
+from celery.result import AsyncResult
+from invenio_db import db
+
+from invenio_webhooks.models import Event, Receiver
+
+
+@shared_task(ignore_results=True)
+def process_event(event_id):
+    """Process event in Celery."""
+    with db.session.begin_nested():
+        event = Event.query.get(event_id)
+        event.receiver.run(event)  # call run directly to avoild circular calls
+        db.session.add(event)
+    db.session.commit()
+
+
+class CeleryReceiver(Receiver):
+    """Asynchronous receiver.
+
+    Receiver which will fire a celery task to handle payload instead of running
+    it synchronously during the request.
+    """
+
+    def __call__(self, event):
+        """Fire a celery task."""
+        process_event.apply_async(args=[event.id])
+
+
+class TaskReceiver(Receiver):
+    """Base class for task-based webhook receivers.
+
+    This receiver runs a long-running task and keep track of its status.
+
+    Payloads sent to this receivers must conform to the following form:
+
+    {
+        "action": "new_task" | "get_status" | "cancel_task"
+        "task_id": <user-defined ID>
+        "kwargs": {
+            "arg1": value1
+            "arg2": value2
+                  .
+                  .
+                  .
+            "argN": valueN
+        }
+    }
+
+    """
+
+    def run(self, event):
+        """Run the requested action."""
+        action = event.payload['action']
```

Comment:
@jirikuncar I added an `Event.action` field and do the branching based on that, so as to avoid "shotgun surgery". WDYT?